internal/accrual/storage/storage.go

```diff
@@ -0,0 +1,245 @@
+package storage
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/jackc/pgerrcode"
+	"github.com/jackc/pgx/v5/pgconn"
+	"github.com/jackc/pgx/v5/pgxpool"
+	"github.com/pressly/goose/v3"
+
+	"github.com/PoorMercymain/gophermart/internal/accrual/domain"
+	"github.com/PoorMercymain/gophermart/internal/accrual/interfaces"
+	"github.com/PoorMercymain/gophermart/pkg/util"
+)
+
+type dbStorage struct {
+	pgxPool *pgxpool.Pool
+	mutex   *sync.Mutex
+}
+
+func NewDBStorage(DSN string) (storage interfaces.Storage, err error) {
+	pg, err := sql.Open("pgx/v5", DSN)
+	if err != nil {
+		util.GetLogger().Infoln(err)
+		return
+	}
+	err = goose.SetDialect("postgres")
+	if err != nil {
+		util.GetLogger().Infoln(err)
+		return
+	}
+
+	err = pg.PingContext(context.Background())
+	if err != nil {
+		util.GetLogger().Infoln(err)
+		return
+	}
+
+	err = goose.Run("up", pg, "./pkg/migrations_accrual")
+	if err != nil {
+		util.GetLogger().Infoln(err)
+		return
+	}
+	pg.Close()
+
+	config, err := pgxpool.ParseConfig(DSN)
+	if err != nil {
+		util.GetLogger().Infoln("Error parsing DSN:", err)
+		return
+	}
+
+	pool, err := pgxpool.NewWithConfig(context.Background(), config)
+	if err != nil {
+		util.GetLogger().Infoln("Error creating pgxpool:", err)
+		return
+	}
+
+	util.GetLogger().Infoln("Pool created", pool, err)
+
+	storage = &dbStorage{pgxPool: pool, mutex: &sync.Mutex{}}
+	return
+}
+
+func (dbs *dbStorage) StoreGoodsReward(ctx context.Context, goods *domain.Goods) (err error) {
+
+	conn, err := dbs.pgxPool.Acquire(ctx)
+	if err != nil {
+		return
+	}
+	defer conn.Release()
+
+	tx, err := conn.Begin(ctx)
+	if err != nil {
+		return
+	}
+	defer tx.Rollback(ctx)
+
+	var pgErr *pgconn.PgError
+
+	currInsertValues := fmt.Sprintf("INSERT INTO goods (id, match, reward, reward_type) "+
+		"VALUES(DEFAULT, '%v', %v, '%v')", goods.Match, goods.Reward, goods.RewardType)
+
+	_, err = tx.Exec(ctx, currInsertValues)
+
+	if err != nil {
+
+		util.GetLogger().Infoln(err)
+		util.GetLogger().Infoln(pgErr)
+		if errors.As(err, &pgErr) && pgErr.Code == pgerrcode.UniqueViolation {
+			tx.Rollback(ctx)
+			util.GetLogger().Infoln(err)
+			return domain.ErrorMatchAlreadyRegistered
+		}
+		return
+	}
+
+	err = tx.Commit(ctx)
+	if err != nil {
+		util.GetLogger().Infoln(err)
+		util.GetLogger().Infoln(pgErr)
+	}
+	return
+}
+func (dbs *dbStorage) StoreOrder(ctx context.Context, order *domain.OrderRecord) (err error) {
+
+	conn, err := dbs.pgxPool.Acquire(ctx)
+	if err != nil {
+		return err
+	}
+	defer conn.Release()
+
+	tx, err := conn.Begin(ctx)
+	if err != nil {
+		return err
+	}
+	defer tx.Rollback(ctx)
+
+	var pgErr *pgconn.PgError
+
+	currInsertValues := fmt.Sprintf("INSERT INTO orders (id, number, status, accrual) VALUES"+
```

Comment:
In SQL, strings must be properly escaped to prevent potential security issues and mishandling of special characters.